pkg/reconciler/pubsub/reconciler.go

```diff
@@ -0,0 +1,174 @@
+}
+
+func (c *PubSubBase) ensureTopicJob(ctx context.Context, action string, owner kmeta.OwnerRefable, project, topic string) (OpsState, error) {
+	job, err := c.getJob(ctx, owner.GetObjectMeta(), labels.SelectorFromSet(operations.TopicJobLabels(owner, action)))
+	// If the resource doesn't exist, we'll create it
+	if apierrs.IsNotFound(err) {
+		c.Logger.Debugw("Job not found, creating.")
+
+		job = operations.NewTopicOps(operations.TopicArgs{
+			Image:     c.TopicOpsImage,
+			Action:    action,
+			ProjectID: project,
+			TopicID:   topic,
+			Owner:     owner,
+		})
+
+		job, err := c.KubeClientSet.BatchV1().Jobs(owner.GetObjectMeta().GetNamespace()).Create(job)
+		if err != nil || job == nil {
+			c.Logger.Debugw("Failed to create Job.", zap.Error(err))
+			return OpsCreateFailedState, nil
+		}
+
+		c.Logger.Debugw("Created Job.")
+		return OpsCreatedState, nil
+	} else if err != nil {
+		c.Logger.Debugw("Failed to get Job.", zap.Error(err))
+		return OpsGetFailedState, err
+	}
+
+	if operations.IsJobComplete(job) {
+		c.Logger.Debugw("Job is complete.")
+		if operations.IsJobSucceeded(job) {
+			return OpsCompeteSuccessfulState, nil
+		} else if operations.IsJobFailed(job) {
+			return OpsCompeteFailedState, errors.New(operations.JobFailedMessage(job))
+		}
+	}
+	c.Logger.Debug("Job still active.", zap.Any("job", job))
+	return OpsOngoingState, nil
+}
+
+func (c *PubSubBase) ensureSubscriptionJob(ctx context.Context, action string, owner kmeta.OwnerRefable, project, topic, subscription string) (OpsState, error) {
+	job, err := c.getJob(ctx, owner.GetObjectMeta(), labels.SelectorFromSet(operations.SubscriptionJobLabels(owner, action)))
+	// If the resource doesn't exist, we'll create it
+	if apierrs.IsNotFound(err) {
+		c.Logger.Debugw("Job not found, creating.")
+
+		job = operations.NewSubscriptionOps(operations.SubArgs{
+			Image:          c.SubscriptionOpsImage,
+			Action:         action,
+			ProjectID:      project,
+			TopicID:        topic,
+			SubscriptionID: subscription,
+			Owner:          owner,
+		})
+
+		job, err := c.KubeClientSet.BatchV1().Jobs(owner.GetObjectMeta().GetNamespace()).Create(job)
+		if err != nil || job == nil {
+			c.Logger.Debugw("Failed to create Job.", zap.Error(err))
+			return OpsCreateFailedState, nil
+		}
+
+		c.Logger.Debugw("Created Job.")
+		return OpsCreatedState, nil
+	} else if err != nil {
+		c.Logger.Debugw("Failed to get Job.", zap.Error(err))
+		return OpsGetFailedState, err
+	}
+
+	if operations.IsJobComplete(job) {
+		c.Logger.Debugw("Job is complete.")
+		if operations.IsJobSucceeded(job) {
+			return OpsCompeteSuccessfulState, nil
+		} else if operations.IsJobFailed(job) {
+			return OpsCompeteFailedState, errors.New(operations.JobFailedMessage(job))
+		}
+	}
+	c.Logger.Debug("Job still active.", zap.Any("job", job))
+	return OpsOngoingState, nil
+}
+
+func (r *PubSubBase) getJob(ctx context.Context, owner metav1.Object, ls labels.Selector) (*v1.Job, error) {
```

Comment:
Is there a reason why we can't use a well defined name and not have to list them using label selectors. I think there might be a race here. We had this same discussion with Channels and decided to use predictable naming and that also has been what serving has been doing.

Reply:
I am using generate name, there should be no race because I am watching the job resource for jobs owned by the resource under reconcile

Reply:
(without looking at any code) There's still a race if work is queued and is processed before the informer cache is updated.

Reply:
@vagababov just created `github.com/knative/serving/pkg/resources.ChildName(parent, suffix string)` for generating suffixed child resource names in a way that automagically shortens the parent name to make the total fit in 63-characters.  I'd be 👍 on moving to `knative/pkg`.

Reply:
followup: https://github.com/GoogleCloudPlatform/cloud-run-events/issues/27